share/wake/lib/core/json.wake

```diff
@@ -169,3 +169,143 @@ export def x ==/ y = match x y
+    normalizeJSON "" json
+
+# The rules by which `normalizeJSONWith` will simplify JSON values.
+# If any particular function is unable to operate on every input value (e.g. a
+# `JObject` contains duplicate keys of types which cannot be combined), that
+# rule may return a `Fail` which halts the broader processing.
+export tuple JSONNormalize =
+    String: String => Result String Error
+    Integer: Integer => Result Integer Error
+    Double: Double => Result Double Error
+    Boolean: Boolean => Result Boolean Error
+    Object: List (Pair String JValue) => Result (List (Pair String JValue)) Error
+    Array: List JValue => Result (List JValue) Error
+
+# A basic normalization ruleset which returns inputs unchanged.
+export def normalizeJSONIdentity: JSONNormalize =
+    JSONNormalize Pass Pass Pass Pass Pass Pass
+
+# Target the minimum JSON language described by the specifications.
+export def normalizeJSONCompat: JSONNormalize =
+    def nfkcKeys =
+        map (editPairFirst unicodeIdentifier)
+    normalizeJSONIdentity
+    | setJSONNormalizeString (Pass _.unicodeCanonical)
+    | setJSONNormalizeDouble filterNonDigitJDouble
+    | setJSONNormalizeObject (deduplicateJObjectKeys _.nfkcKeys)
+
+# Fail on any `Double` values which can't be represented numerically.
+# Published JSON specifications describe their double values as allowing decimal
+# or exponential forms, but don't implement the full IEEE standard.
+def filterNonDigitJDouble (n: Double): Result Double Error = match n.dclass
+    DoubleInfinite =
+        failWithError "For compatibility, JSON doubles should not be infinite."
+    DoubleNaN =
+        failWithError "For compatibility, JSON doubles should not be NaN values."
+    _ =
+        Pass n
+
+# Attempt to simplify any `JObject`s which contain multiple instances of a key.
+# While the published specifications explicitly allow such duplicate keys, none
+# *require* implementations to accept them nor describes the behaviour by which
+# they should be handled; therefore, implementations differ in how they handle
+# duplicate keys -- if they do so at all -- and a generator aiming for
+# compatibility shouldn't output objects with duplicate keys.
+#
+# In particular, this function will concatenate the members of two `JArray`
+# values, treat `JNull` as a wildcard placeholder, and otherwise return the
+# original value if both instances are equal. It fails if a key is associated
+# with multiple values of different types or with different values of a type
+# which cannot be combined.
+def deduplicateJObjectKeys (dict: List (Pair String JValue)): Result (List (Pair String JValue)) Error =
+    def cmpKeysOnly (Pair k1 _) (Pair k2 _) =
+        k1 <* k2
+    def simplifyKeyGroups = match _
+        Nil =
+            # `simplifyKeyGroups` is only ever mapped over the output of
+            # `groupBy`, which is always a list of *non-empty* lists.
+            failWithError "groupBy invariant reached in deduplicateJObjectKeys"
```

Comment:
> This one should probably be `unreachable`.

I'm not a fan of `unreachable` on principle. Since the return type already comes with a built-in `Error`, this can easily slot in to that infrastructure and never be run either way.

Reply:
With `unreachable` actually being a compiler hint, I agree that makes more sense here.  I'll change it over.